### PR TITLE
On demand property derivation (scalar)

### PR DESCRIPTION
**Pull request by chrishajas**
This is the follow up PR to https://github.com/greenplum-db/gporca/pull/511. It uses the infrastructure built in that PR and follows the same pattern to derive scalar properties on demand. Commits are broken up in similar sections to that PR for ease of reviewing.

### CODE REVIEW

File: libgpopt/src/operators/CExpressionHandle.cpp [hunk 1]
```diff
@@ -1437,7 +1436,7 @@ CExpressionHandle::PexprScalarChild
 		GPOS_ASSERT((*m_pgexpr)[child_index]->FScalar());
 
 		CExpression *pexprScalar = (*m_pgexpr)[child_index]->PexprScalar();
-		GPOS_ASSERT_IMP(NULL == pexprScalar, CDrvdPropScalar::GetDrvdScalarProps((*m_pgexpr)[child_index]->Pdp())->FHasSubquery());
+		GPOS_ASSERT_IMP(NULL == pexprScalar, CDrvdPropScalar::GetDrvdScalarProps((*m_pgexpr)[child_index]->Pdp())->GetHasSubquery());
```
Comment (ashuka24) [inline libgpopt/src/operators/CExpressionHandle.cpp]:
``` suggestion
		GPOS_ASSERT_IMP(NULL == pexprScalar, (*m_pgexpr)[child_index]->DeriveHasSubquery());
```

File: libgpopt/include/gpopt/base/CDrvdPropScalar.h [hunk 2]
```diff
@@ -94,73 +147,45 @@ namespace gpopt
...
-				return m_pcrsSetReturningFunction;
-			}
+			CColRefSet *GetSetReturningFunctionColumns() const;
 
 			// do subqueries appear in the operator's tree?
-			BOOL FHasSubquery() const
-			{
-				return m_fHasSubquery;
-			}
+			BOOL GetHasSubquery() const;
```
Comment (ashuka24) [inline libgpopt/include/gpopt/base/CDrvdPropScalar.h]: Can this just be `HasSubquery`? GetHas sounds kind of strange.

File: libgpopt/src/base/COptimizationContext.cpp [hunk 3]
```diff
@@ -318,7 +318,7 @@ COptimizationContext::FOptimizeAgg
 
 	if (GPOS_FTRACE(EopttraceForceExpandedMDQAs))
 	{
-		BOOL fHasMultipleDistinctAggs = CDrvdPropScalar::GetDrvdScalarProps((*pgexprAgg)[1]->Pdp())->FHasMultipleDistinctAggs();
+		BOOL fHasMultipleDistinctAggs = CDrvdPropScalar::GetDrvdScalarProps((*pgexprAgg)[1]->Pdp())->GetHasMultipleDistinctAggs();
```
Comment (ashuka24) [inline libgpopt/src/base/COptimizationContext.cpp]:
```suggestion
		BOOL fHasMultipleDistinctAggs = (*pgexprAgg)[1]->DeriveHasMultipleDistinctAggs();
```

Comment (ashuka24) [reply libgpopt/src/base/COptimizationContext.cpp]: This actually seems like it was intentionally left. Could you make it clear somewhere that we cannot do on demand property derivation for group expressions?

File: libgpopt/src/operators/CExpressionHandle.cpp (hunk 1 above)
Comment (ashuka24) [reply libgpopt/src/operators/CExpressionHandle.cpp]: This actually seems like it was intentionally left. Could you make it clear somewhere that we cannot do on demand property derivation for group expressions?

File: libgpopt/src/base/CConstraintInterval.cpp [hunk 4]
```diff
@@ -162,8 +162,7 @@ CConstraintInterval::PciIntervalFromScalarExpr
 
 	// expression must use at most one column
 #ifdef GPOS_DEBUG
```
Comment (zellerh) [inline libgpopt/src/base/CConstraintInterval.cpp]: Could now also remove the #ifdefs

File: libgpopt/src/base/CDrvdPropScalar.cpp [hunk 5]
```diff
@@ -196,68 +162,237 @@ CDrvdPropScalar::PcrsDefined() const
...
+				m_pcrsUsed->Union(exprhdl.DeriveOuterReferences(0));
+			}
+		}
+	}
+	return m_pcrsUsed;
+}
+
 // columns containing set-returning function
 CColRefSet *
 CDrvdPropScalar::PcrsSetReturningFunction() const
```
[the remark this answers is not in the document]
Comment (zellerh) [reply libgpopt/src/base/CDrvdPropScalar.cpp]: Never mind, you already did that.

File: libgpopt/include/gpopt/base/CDrvdPropScalar.h [hunk 6]
```diff
@@ -78,10 +97,44 @@ namespace gpopt
...
 			CDrvdPropScalar(const CDrvdPropScalar &);
 
+			// Have all the properties been derivied?
+			//
+			// NOTE1: This is set ONLY when Derive() is called. If all the properties
+			// are independently derived, m_is_complete will remain false. In that
+			// case, even though Derive() would attempt to derive all the properties
+			// once again, it should be quick, since each individual member has been
+			// cached.
+			// NOTE2: Once these properties are detached from the
```
[the remark this answers is not in the document]
Comment (chrishajas) [reply libgpopt/include/gpopt/base/CDrvdPropScalar.h]: Whenever we detach the properties from the expression, which we do by calling a function that returns CDrvdProp, we derive all scalar or relational properties. So no, when we detach the properties, we derive everything, since as you mentioned they could be used in the future.

File: libgpopt/src/operators/CExpressionHandle.cpp (hunk 1 above)
Comment (chrishajas) [reply libgpopt/src/operators/CExpressionHandle.cpp]: Added a comment in `CExpressionHandle` above the `DeriveXXX` functions. Also, for group expressions there is no associated `DeriveXXX` function that will compile.

File: libgpopt/src/operators/CExpressionHandle.cpp [hunk 7]
```diff
@@ -1764,6 +1748,10 @@ CExpressionHandle::Pstats()
 	return m_pstats;
 }
 
+// The below functions use on-demand property derivation
+// only if there is an expression associated with the expression handle.
+// If there is only a group expression or a cost context assoicated with the handle,
+// all properties must have already been derived as we can't derive anything.
```
Comment (ashuka24) [inline libgpopt/src/operators/CExpressionHandle.cpp]: remove extra line

File: libgpopt/src/operators/CExpressionHandle.cpp [hunk 8]
```diff
@@ -1049,23 +1049,17 @@ CExpressionHandle::GetRelationalProperties
 			return (*Pexpr())[child_index]->GetDrvdPropRelational();
 		}
 
-		// return props after calling derivation function. This calls the on-demand function of the expression
+		// return props after calling derivation function
 		return CDrvdPropRelational::GetRelationalProperties((*Pexpr())[child_index]->PdpDerive());
 	}
 
-	if (FAttachedToLeafPattern())
```
Comment (ashuka24) [inline libgpopt/src/operators/CExpressionHandle.cpp]: why was this removed?

Comment (chrishajas) [reply libgpopt/src/operators/CExpressionHandle.cpp]: It's an unreachable condititon. FAttachedToLeafPattern() is doing 
```
return
		0 == Arity() &&
		NULL != m_pexpr &&
		NULL != m_pexpr->Pgexpr();
```
which can't be reached since we check `NULL != m_pexpr` in the if statement above it.

File: libgpopt/include/gpopt/base/CDrvdPropScalar.h [hunk 9]
```diff
@@ -40,9 +40,27 @@ namespace gpopt
...
+			EdptFHasMultipleDistinctAggs,
+			EdptFHasScalarArrayCmp,
+			EdptSentinel
+		};
 
 		private:
 
+			CMemoryPool *m_mp;
+
+			CBitSet *m_is_prop_derived;
```
[the remark this answers is not in the document]
Comment (chrishajas) [reply libgpopt/include/gpopt/base/CDrvdPropScalar.h]: Agree, it's something that will give us benefit in quite a few places.
